# Lab: Build your first image encoder–decoder

**Time:** 60–90 minutes. **Prerequisites:** Python, tensors, basic convolutions, and gradient descent.

Read [CNN-Based Encoder–Decoder Architectures for Images](../../lectures/encoder_decoder/encoder_decoder.md), especially Sections 2–7.

You will reconstruct handwritten digits using a small convolutional network. Each image is both the input and the target: this training task makes our encoder–decoder an **autoencoder**. Digit labels are not used for training.

By the end, you should be able to:
- implement an encoder, spatial latent representation, and decoder;
- trace `[N, C, H, W]` shapes through downsampling and upsampling;
- train with reconstruction MSE and inspect held-out reconstructions;
- measure latent size and explain the bottleneck trade-off.

Complete the four TODO sections. Data loading, plotting, and evaluation are provided. TODO cells intentionally raise `NotImplementedError` until you implement them. Run cells in order.

## 1. Set up and load data (10 minutes)
Use a Python 3 notebook kernel. If needed, uncomment the installation cell, run it once, and restart the kernel. A GPU is optional. The first dataset load needs internet access; later runs use the local cache. `ToTensor()` scales MNIST pixels to `[0, 1]`; do not apply mean/std normalization for this lab.

In [1]:
# %pip install torch torchvision matplotlib

In [2]:
import torch
from torch import nn
from torch.utils.data import DataLoader, Subset
from torchvision import datasets, transforms
import matplotlib.pyplot as plt

torch.manual_seed(42)
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
BATCH_SIZE = 64
EPOCHS = 5  # Start with 1 to check your code, then train for 5.
LATENT_CHANNELS = 8
print("Device:", device)

train_data = datasets.MNIST("./data", train=True, download=True, transform=transforms.ToTensor())
test_data = datasets.MNIST("./data", train=False, download=True, transform=transforms.ToTensor())
# Fixed random subsets keep the exercise small and comparisons repeatable.
g = torch.Generator().manual_seed(42)
train_data = Subset(train_data, torch.randperm(len(train_data), generator=g)[:5000].tolist())
test_data = Subset(test_data, torch.randperm(len(test_data), generator=g)[:1000].tolist())
train_loader = DataLoader(train_data, batch_size=BATCH_SIZE, shuffle=True, num_workers=0)
test_loader = DataLoader(test_data, batch_size=BATCH_SIZE, shuffle=False, num_workers=0)
images, labels = next(iter(train_loader))
print("Batch:", tuple(images.shape), "Pixel range:", images.min().item(), images.max().item())
fig, axes = plt.subplots(1, 8, figsize=(12, 2))
for ax, img in zip(axes, images[:8]):
    ax.imshow(img[0], cmap="gray", vmin=0, vmax=1)
    ax.axis("off")
plt.show()

Device: cpu


  0%|          | 0.00/9.91M [00:00<?, ?B/s]

  0%|          | 32.8k/9.91M [00:00<01:55, 85.8kB/s]

  1%|          | 65.5k/9.91M [00:00<01:33, 105kB/s] 

  2%|▏         | 164k/9.91M [00:00<00:39, 244kB/s] 

  3%|▎         | 295k/9.91M [00:01<00:26, 366kB/s]

  5%|▌         | 524k/9.91M [00:01<00:19, 477kB/s]

 13%|█▎        | 1.28M/9.91M [00:01<00:05, 1.50MB/s]

 17%|█▋        | 1.64M/9.91M [00:01<00:04, 1.82MB/s]

 20%|██        | 2.00M/9.91M [00:01<00:05, 1.53MB/s]

 22%|██▏       | 2.23M/9.91M [00:02<00:07, 983kB/s] 

 24%|██▍       | 2.42M/9.91M [00:03<00:10, 741kB/s]

 27%|██▋       | 2.65M/9.91M [00:03<00:08, 881kB/s]

 28%|██▊       | 2.82M/9.91M [00:03<00:08, 808kB/s]

 30%|██▉       | 2.95M/9.91M [00:03<00:09, 739kB/s]

 31%|███       | 3.08M/9.91M [00:04<00:12, 548kB/s]

 32%|███▏      | 3.18M/9.91M [00:04<00:12, 520kB/s]

 33%|███▎      | 3.31M/9.91M [00:04<00:12, 536kB/s]

 35%|███▍      | 3.44M/9.91M [00:04<00:11, 546kB/s]

 36%|███▌      | 3.54M/9.91M [00:05<00:12, 490kB/s]

 36%|███▋      | 3.60M/9.91M [00:05<00:13, 457kB/s]

 37%|███▋      | 3.70M/9.91M [00:05<00:13, 448kB/s]

 38%|███▊      | 3.80M/9.91M [00:05<00:13, 443kB/s]

 39%|███▉      | 3.90M/9.91M [00:05<00:14, 426kB/s]

 40%|████      | 4.00M/9.91M [00:06<00:18, 323kB/s]

 47%|████▋     | 4.65M/9.91M [00:06<00:04, 1.07MB/s]

 49%|████▉     | 4.88M/9.91M [00:06<00:05, 1.01MB/s]

 52%|█████▏    | 5.14M/9.91M [00:07<00:04, 1.02MB/s]

 56%|█████▌    | 5.51M/9.91M [00:07<00:03, 1.14MB/s]

 59%|█████▊    | 5.80M/9.91M [00:07<00:03, 1.17MB/s]

 61%|██████▏   | 6.09M/9.91M [00:07<00:03, 1.19MB/s]

 65%|██████▌   | 6.49M/9.91M [00:08<00:02, 1.32MB/s]

 69%|██████▉   | 6.85M/9.91M [00:08<00:02, 1.46MB/s]

 73%|███████▎  | 7.27M/9.91M [00:08<00:01, 1.53MB/s]

 78%|███████▊  | 7.70M/9.91M [00:08<00:01, 1.93MB/s]

 80%|████████  | 7.96M/9.91M [00:08<00:01, 1.75MB/s]

 83%|████████▎ | 8.26M/9.91M [00:08<00:00, 1.81MB/s]

 87%|████████▋ | 8.59M/9.91M [00:09<00:00, 2.08MB/s]

 89%|████████▉ | 8.85M/9.91M [00:09<00:00, 1.86MB/s]

 95%|█████████▍| 9.40M/9.91M [00:09<00:00, 2.30MB/s]

 98%|█████████▊| 9.67M/9.91M [00:09<00:00, 2.28MB/s]

100%|██████████| 9.91M/9.91M [00:09<00:00, 1.04MB/s]

  0%|          | 0.00/28.9k [00:00<?, ?B/s]

100%|██████████| 28.9k/28.9k [00:00<00:00, 32.1kB/s]

100%|██████████| 28.9k/28.9k [00:00<00:00, 32.0kB/s]

  0%|          | 0.00/1.65M [00:00<?, ?B/s]

  2%|▏         | 32.8k/1.65M [00:01<01:10, 22.9kB/s]

  4%|▍         | 65.5k/1.65M [00:03<01:37, 16.3kB/s]

  6%|▌         | 98.3k/1.65M [00:08<02:34, 10.0kB/s]

  8%|▊         | 131k/1.65M [00:10<02:07, 11.9kB/s] 

 10%|▉         | 164k/1.65M [00:14<02:28, 10.0kB/s]

 12%|█▏        | 197k/1.65M [00:18<02:29, 9.72kB/s]

 14%|█▍        | 229k/1.65M [00:22<02:40, 8.85kB/s]

 16%|█▌        | 262k/1.65M [00:26<02:32, 9.11kB/s]

 18%|█▊        | 295k/1.65M [00:29<02:23, 9.44kB/s]

 20%|█▉        | 328k/1.65M [00:33<02:26, 9.01kB/s]

 22%|██▏       | 360k/1.65M [00:35<02:03, 10.4kB/s]

 24%|██▍       | 393k/1.65M [00:37<01:50, 11.4kB/s]

 26%|██▌       | 426k/1.65M [00:40<01:44, 11.7kB/s]

 28%|██▊       | 459k/1.65M [00:44<02:00, 9.86kB/s]

 30%|██▉       | 492k/1.65M [00:47<01:53, 10.2kB/s]

 32%|███▏      | 524k/1.65M [00:52<02:09, 8.70kB/s]

 34%|███▍      | 557k/1.65M [00:56<02:06, 8.62kB/s]

 36%|███▌      | 590k/1.65M [01:00<02:00, 8.76kB/s]

 38%|███▊      | 623k/1.65M [01:04<02:06, 8.13kB/s]

 40%|███▉      | 655k/1.65M [01:09<02:06, 7.83kB/s]

 42%|████▏     | 688k/1.65M [01:13<01:58, 8.08kB/s]

 44%|████▎     | 721k/1.65M [01:18<02:02, 7.55kB/s]

 46%|████▌     | 754k/1.65M [01:22<01:58, 7.57kB/s]

 48%|████▊     | 786k/1.65M [01:25<01:46, 8.08kB/s]

 50%|████▉     | 819k/1.65M [01:28<01:27, 9.51kB/s]

 52%|█████▏    | 852k/1.65M [01:32<01:32, 8.59kB/s]

 54%|█████▎    | 885k/1.65M [01:37<01:37, 7.83kB/s]

 56%|█████▌    | 918k/1.65M [01:39<01:19, 9.20kB/s]

 58%|█████▊    | 950k/1.65M [01:41<01:03, 11.1kB/s]

 60%|█████▉    | 983k/1.65M [01:43<00:55, 11.9kB/s]

 62%|██████▏   | 1.02M/1.65M [01:45<00:49, 12.9kB/s]

 64%|██████▎   | 1.05M/1.65M [01:48<00:45, 13.1kB/s]

 66%|██████▌   | 1.08M/1.65M [01:53<00:57, 9.87kB/s]

 68%|██████▊   | 1.11M/1.65M [01:56<00:51, 10.5kB/s]

 70%|██████▉   | 1.15M/1.65M [01:58<00:45, 10.9kB/s]

 72%|███████▏  | 1.18M/1.65M [02:00<00:38, 12.2kB/s]

 74%|███████▎  | 1.21M/1.65M [02:02<00:31, 13.9kB/s]

 76%|███████▌  | 1.25M/1.65M [02:02<00:22, 18.2kB/s]

 78%|███████▊  | 1.28M/1.65M [02:05<00:22, 16.5kB/s]

 79%|███████▉  | 1.31M/1.65M [02:08<00:23, 14.2kB/s]

 81%|████████▏ | 1.34M/1.65M [02:10<00:20, 15.1kB/s]

 83%|████████▎ | 1.38M/1.65M [02:11<00:15, 18.1kB/s]

 85%|████████▌ | 1.41M/1.65M [02:12<00:12, 19.9kB/s]

 87%|████████▋ | 1.44M/1.65M [02:14<00:11, 18.1kB/s]

 89%|████████▉ | 1.47M/1.65M [02:17<00:12, 14.2kB/s]

 91%|█████████▏| 1.51M/1.65M [02:19<00:08, 16.8kB/s]

 93%|█████████▎| 1.54M/1.65M [02:22<00:07, 13.8kB/s]

 95%|█████████▌| 1.57M/1.65M [02:25<00:05, 13.0kB/s]

 97%|█████████▋| 1.61M/1.65M [02:28<00:03, 11.9kB/s]

 99%|█████████▉| 1.64M/1.65M [02:31<00:00, 11.7kB/s]

100%|██████████| 1.65M/1.65M [02:31<00:00, 10.9kB/s]

  0%|          | 0.00/4.54k [00:00<?, ?B/s]

100%|██████████| 4.54k/4.54k [00:00<00:00, 556kB/s]

Batch: (64, 1, 28, 28) Pixel range: 0.0 1.0


C:\Users\pooja\AppData\Local\Temp\ipykernel_18192\2172445123.py:28: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


## 2. Predict the shapes (10 minutes)
Before coding, fill in the blanks. Use the convolution formulas in Sections 3.1 and 5.1 of the lecture. Assume dilation 1.

| Stage | Layer | Output shape |
|---|---|---|
| Input | — | `[N, 1, 28, 28]` |
| Encoder 1 | Conv2d: 1 → 16, kernel 3, stride 2, padding 1; ReLU | `[N, 16, 14, 14]` |
| Encoder 2 | Conv2d: 16 → 32, kernel 3, stride 2, padding 1; ReLU | `[N, 32, 7, 7]` |
| Bottleneck | Conv2d: 32 → L, kernel 1 | `[N, L, 7, 7]` |
| Decoder 1 | ConvTranspose2d: L → 16, kernel 3, stride 2, padding 1, output_padding 1; ReLU | `[N, 16, 14, 14]` |
| Decoder 2 | ConvTranspose2d: 16 → 1, same spatial settings; Sigmoid | `[N, 1, 28, 28]` |

**Write your answers here:**
1. For `L=8`, input values per image = 784; latent values = 392; input/latent ratio = 2.
2. For `L=32`, no. Latent values are \(32 \times 7 \times 7 = 1568\), which is more than the 784 input pixels. The spatial grid is smaller, but the tensor is larger than the input.
3. `ToTensor()` scales pixels to `[0, 1]`. Sigmoid keeps the reconstruction in that same range, so MSE compares values on the same scale.
4. With `output_padding=0`, the first decoder height is \((7-1)\times 2 - 2\times 1 + (3-1) + 1 = 13\).

The 1×1 bottleneck mixes channels at each location. It preserves the spatial grid; no flattening or linear layer is needed. The input/latent ratio counts tensor values, not file size or actual encoded bits.

## 3. Implement the network (20 minutes)
**TODO 1:** Define `self.network` in `Encoder` using the three convolutional layers in the table and ReLU after the first two. Leave the bottleneck output without an activation.

**TODO 2:** Define `self.network` in `Decoder` using the two transposed convolutions, ReLU between them, and Sigmoid at the end.

**TODO 3:** Implement `forward` in the combined model. Return `(reconstruction, z)`. Hints: use `nn.Sequential`, `nn.Conv2d`, and `nn.ConvTranspose2d`.

In [3]:
class Encoder(nn.Module):
    def __init__(self, latent_channels=8):
        super().__init__()
        self.network = nn.Sequential(
            nn.Conv2d(1, 16, kernel_size=3, stride=2, padding=1),
            nn.ReLU(),
            nn.Conv2d(16, 32, kernel_size=3, stride=2, padding=1),
            nn.ReLU(),
            nn.Conv2d(32, latent_channels, kernel_size=1),
        )

    def forward(self, x):
        return self.network(x)

In [4]:
class Decoder(nn.Module):
    def __init__(self, latent_channels=8):
        super().__init__()
        self.network = nn.Sequential(
            nn.ConvTranspose2d(
                latent_channels, 16, kernel_size=3, stride=2, padding=1, output_padding=1
            ),
            nn.ReLU(),
            nn.ConvTranspose2d(16, 1, kernel_size=3, stride=2, padding=1, output_padding=1),
            nn.Sigmoid(),
        )

    def forward(self, z):
        return self.network(z)

In [5]:
class ImageEncoderDecoder(nn.Module):
    def __init__(self, latent_channels=8):
        super().__init__()
        self.encoder = Encoder(latent_channels)
        self.decoder = Decoder(latent_channels)

    def forward(self, x):
        z = self.encoder(x)
        reconstruction = self.decoder(z)
        return reconstruction, z

In [6]:
# Shape and range checks: run these before training.
model = ImageEncoderDecoder(LATENT_CHANNELS).to(device)
x = torch.rand(4, 1, 28, 28, device=device)
with torch.no_grad():
    reconstruction, z = model(x)
assert z.shape == (4, LATENT_CHANNELS, 7, 7), "Check encoder stride/padding/channels"
assert reconstruction.shape == x.shape, "Check decoder output_padding"
assert torch.isfinite(reconstruction).all(), "Output contains nonfinite values"
assert ((reconstruction >= 0) & (reconstruction <= 1)).all(), "Check output activation"
print("Input:", tuple(x.shape), "Latent:", tuple(z.shape), "Output:", tuple(reconstruction.shape))
print("Shape and range checks passed")

Input: (4, 1, 28, 28) Latent: (4, 8, 7, 7) Output: (4, 1, 28, 28)
Shape and range checks passed


## 4. Train to reconstruct (15 minutes)
**TODO 4:** Complete one training step. Clear old gradients, run the model, compute MSE against `images`, backpropagate, and update the weights. Return the loss as a Python number with `.item()`.

Both components learn together because the optimizer receives all model parameters. The provided evaluation code uses held-out images, evaluation mode, and no gradient tracking. MSE is averaged over all pixels and images; it is not classification accuracy.

In [7]:
def train_step(model, images, optimizer, criterion):
    optimizer.zero_grad()
    reconstruction, _ = model(images)
    loss = criterion(reconstruction, images)
    loss.backward()
    optimizer.step()
    return loss.item()

In [8]:
criterion = nn.MSELoss()

@torch.no_grad()
def evaluate(model, loader):
    model.eval()
    total_loss = 0.0
    for images, _ in loader:
        images = images.to(device)
        reconstruction, _ = model(images)
        total_loss += criterion(reconstruction, images).item() * images.size(0)
    return total_loss / len(loader.dataset)

# Re-running this cell starts a fresh model and optimizer.
torch.manual_seed(42)
model = ImageEncoderDecoder(LATENT_CHANNELS).to(device)
optimizer = torch.optim.Adam(model.parameters(), lr=1e-3)
initial_mse = evaluate(model, test_loader)
print(f"Untrained held-out MSE: {initial_mse:.5f}")
history = {"train": [], "test": []}
for epoch in range(EPOCHS):
    model.train()
    total_loss = 0.0
    for images, _ in train_loader:
        images = images.to(device)
        total_loss += train_step(model, images, optimizer, criterion) * images.size(0)
    history["train"].append(total_loss / len(train_loader.dataset))
    history["test"].append(evaluate(model, test_loader))
    print(f"Epoch {epoch + 1}: train MSE={history['train'][-1]:.5f}, "
          f"held-out MSE={history['test'][-1]:.5f}")

plt.plot(range(1, EPOCHS + 1), history["train"], marker="o", label="Train")
plt.plot(range(1, EPOCHS + 1), history["test"], marker="o", label="Held-out")
plt.xlabel("Epoch")
plt.ylabel("Mean squared error")
plt.legend()
plt.show()

Untrained held-out MSE: 0.18629


Epoch 1: train MSE=0.11827, held-out MSE=0.04370


Epoch 2: train MSE=0.03428, held-out MSE=0.02552


Epoch 3: train MSE=0.01682, held-out MSE=0.00983


Epoch 4: train MSE=0.00827, held-out MSE=0.00620


Epoch 5: train MSE=0.00572, held-out MSE=0.00460


C:\Users\pooja\AppData\Local\Temp\ipykernel_18192\2217293183.py:36: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


## 5. Inspect the reconstruction and latent (10 minutes)
Look for digit identity, stroke thickness, and missing detail. A low average pixel error alone does not guarantee sharp or useful images. Feature maps are learned activations, not necessarily human-readable parts of a digit.

In [9]:
model.eval()
examples, _ = next(iter(test_loader))
with torch.no_grad():
    reconstructed, latents = model(examples[:8].to(device))
reconstructed, latents = reconstructed.cpu(), latents.cpu()
fig, axes = plt.subplots(2, 8, figsize=(12, 4))
for i in range(8):
    axes[0, i].imshow(examples[i, 0], cmap="gray", vmin=0, vmax=1)
    axes[1, i].imshow(reconstructed[i, 0], cmap="gray", vmin=0, vmax=1)
    axes[0, i].axis("off")
    axes[1, i].axis("off")
fig.suptitle("Top: original | Bottom: reconstruction")
plt.show()

count = min(LATENT_CHANNELS, 8)
fig, axes = plt.subplots(1, count, figsize=(2 * count, 2), squeeze=False)
for i in range(count):
    axes[0, i].imshow(latents[0, i], cmap="viridis")
    axes[0, i].set_title(f"Channel {i}")
    axes[0, i].axis("off")
fig.suptitle("Latent maps of the first image (each map scaled separately)")
plt.show()

C:\Users\pooja\AppData\Local\Temp\ipykernel_18192\1380744808.py:13: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


C:\Users\pooja\AppData\Local\Temp\ipykernel_18192\1380744808.py:22: UserWarning: FigureCanvasAgg is non-interactive, and thus cannot be shown
  plt.show()


## 6. Experiment with the bottleneck (10–15 minutes)
Run the lab with `LATENT_CHANNELS = 2`, `8`, and `32`. Change the value in the setup cell and rerun from that cell downward. Keep the data, seed, learning rate, and epoch count fixed. The training cell creates a fresh model each time. Save your observations before the next run.

| Latent channels | Values per image | Input/latent ratio | Final train MSE | Final held-out MSE | Visible differences |
|---|---|---|---|---|---|
| 2 | 98 | 8 | 0.01209 | 0.01069 | Digits are recognizable but soft. Thin strokes thicken, and the last 9 loses its hook. |
| 8 | 392 | 2 | 0.00572 | 0.00460 | The executed run in this notebook. Digits are clearer, with some remaining blur. |
| 32 | 1568 | 0.5 | 0.00350 | 0.00295 | Strokes are sharper and closer to the originals, including the last 9. |

This is an exploratory comparison: changing channels also changes the parameter count. Small runs may not show a consistent ranking. Since we repeatedly inspect this held-out subset, treat it as validation data rather than an untouched final benchmark.

**Discuss:**
1. Did a larger latent improve reconstruction in your runs? Support the answer with images and MSE.
2. Why is spatial downsampling alone insufficient to claim dimensional compression?
3. Why can the decoder restore image size without recovering every original detail?
4. How would a `[N, 128]` vector differ from this spatial latent?
5. Where could a skip connection carry high-resolution features? How might that weaken the bottleneck constraint?
6. Why does this exercise reconstruct existing images rather than establish a reliable random-image generator?

**Answers:**
1. Yes. Held-out MSE fell from 0.01069 at `L=2` to 0.00460 at `L=8` and 0.00295 at `L=32`. The images match that: `L=2` blurs thin strokes and drops the hook of a 9, while `L=32` keeps stroke shape. The parameter count also changes, so this is not a pure bottleneck comparison.
2. Every setting downsamples to `7×7`. At `L=32` the latent still has \(32\times7\times7=1568\) values, more than the 784 input pixels, so the spatial shrink is not compression.
3. Transposed convolutions only restore the `28×28` canvas. Detail discarded at the bottleneck has no unique way back, so the decoder draws a plausible digit rather than the original strokes.
4. A `[N, 128]` vector has no height or width. This latent is `[N, L, 7, 7]`, so each channel is still a coarse map and nearby image regions stay roughly in place.
5. A skip could copy the encoder's `14×14` maps, or the `28×28` input, into the matching decoder stage. The decoder could then paste fine detail around the bottleneck, so `z` would constrain the reconstruction less.
6. Training only learns \(x \rightarrow z \rightarrow \hat{x}\) on real digits. `z` is not forced to come from a known distribution, so a random code is not a new digit. This model reconstructs images it has seen; it does not define a generator.

## Optional extension: denoising
Keep the architecture and start with a fresh model and optimizer. Add noise to the training **input**, while retaining the clean image as the **target**:

```python
noisy = (images + 0.3 * torch.randn_like(images)).clamp(0, 1)
reconstruction, _ = model(noisy)
loss = criterion(reconstruction, images)
```

Update both training and evaluation to use noisy inputs and clean targets. For evaluation, generate one fixed set of noisy held-out images and reuse it. Compare noisy-input MSE with denoised-output MSE against the clean images, and plot clean/noisy/reconstructed rows. Merely feeding noise to a reconstruction-trained model is not the same as training a denoiser.

An alternative extension is to replace each transposed convolution with 2× bilinear upsampling (`align_corners=False`) followed by a 3×3 convolution with padding 1. Keep the intermediate ReLU and final Sigmoid and check shapes before retraining.

## Submission and assessment
Submit the completed notebook with:
- all four TODOs implemented and shape checks passing (**4 marks**);
- your completed shape table and latent-size calculations (**2 marks**);
- training curves and original/reconstructed images (**2 marks**);
- the three-run experiment table and evidence-based discussion (**2 marks**).

Expected behavior: reconstruction error should generally decrease and reconstructed digits should become recognizable, though some strokes may remain blurred. No fixed MSE threshold is required. If training fails, check the target, output range, gradient update, and tensor shapes before increasing epochs.